### PR TITLE
[incubator/consul] Add new Consul chart

### PR DESCRIPTION
**Pull request by tompizmor**
List with things tested and pending ones:

- [x] Upscaling/Downscaling the cluster from 3 to 5 nodes and 5 to 3 (those numbers are the recommended by the Consul team)
- [x] Secret with the private key it's taken by the nami module and set in the right place
- [x] Values parametrized in values.yaml file
- [x] Load configuration via env var
- [x] Load configuration via configMap
- [x] Ingress to access the user interface
- [x] Node affinity settings
- [x] README
- [x]  Review notes.txt
- [x]  Prometheus Consul exporter
- [x]  Check structure of values.yaml. (image part)
- [x]  Secret of TLS for ingress (Documentation)
- [x]  Secret for TLS between servers (secret structure and documentation)
- [x] Check the logic of the Nami module for the persistence of the configuration files outside k8s.

### CODE REVIEW

File: incubator/consul/templates/NOTES.txt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    export SERVICE_IP=$(kubectl get svc --namespace {{ .Release.Namespace }} {{ template "consul.fullname" . }}-ui -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
+
+    echo http://$SERVICE_IP:{{ .Values.httpPort }}/
+
+
+{{- else if contains "ClusterIP"  .Values.uiService.type }}
+
+    kubectl port-forward {{ template "consul.fullname" . }}-0 --namespace {{ .Release.Namespace }} {{ .Values.httpPort }}:{{ .Values.httpPort }}
+
+    echo URL : http://127.0.0.1:{{ .Values.httpPort }}
```
[the remark this answers is not in the document]
Comment (tompizmor) [reply incubator/consul/templates/NOTES.txt]: Tt is just for display the info. The user will copy paste the command and see this:

```
URL : http://127.0.0.1:8500
```

[the remark this answers is not in the document]
Comment (juan131) [reply incubator/consul/templates/NOTES.txt]: That's not true... See the result:
```
In order to access to the Consul Web UI:

    kubectl port-forward consul-test-consul-0 --namespace default 8500:8500

    echo URL : http://127.0.0.1:8500
```

[the remark this answers is not in the document]
Comment (juan131) [reply incubator/consul/templates/NOTES.txt]: That's why I suggested using:

```
curl http://127.0.0.1:8500
```

File: incubator/consul/templates/consul-configmap.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+{{ if .Values.configmap }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ template "consul.fullname" . }}-configuration
+  labels:
+    app: {{ template "consul.name" . }}
+    chart: {{ template "consul.chart" . }}
+    release: "{{ .Release.Name }}"
```
[the remark this answers is not in the document]
Comment (tompizmor) [reply incubator/consul/templates/consul-configmap.yaml]: 👍

File: incubator/consul/templates/pod-dist-budget.yaml [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    chart: {{ template "consul.chart" . }}
+    release: "{{ .Release.Name }}"
+    heritage: "{{ .Release.Service }}"
+spec:
+  maxUnavailable: {{ .Values.maxUnavailable }}
+  selector:
+    matchLabels:
+      heritage: {{ .Release.Service | quote }}
+      release: {{ .Release.Name | quote }}
+      chart: "{{ .Chart.Name }}-{{ .Chart.Version }}"
```
[the remark this answers is not in the document]
Comment (tompizmor) [reply incubator/consul/templates/pod-dist-budget.yaml]: 👍

File: incubator/consul/templates/NOTES.txt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+{{- end }}
+
+Please take into account that you need to wait until a cluster leader is elected. 
+In order to check the status of the cluster you can run the following command:
+
+    kubectl exec -it {{ template "consul.fullname" . }}-0 -- consul operator raft list-peers
```
Comment (juan131) [inline incubator/consul/templates/NOTES.txt]: Can I suggest adding one more command to check the status of the cluster?

```
kubectl exec -it {{ template "consul.fullname" . }}-0 -- consul members
```

File: incubator/consul/templates/_helpers.tpl [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+{{- $name := default .Chart.Name .Values.nameOverride -}}
+{{- printf "%s-%s" .Release.Name $name | trunc 63 | trimSuffix "-" -}}
+{{- end -}}
+
+{{/*
+Create chart name and version as used by the chart label.
+*/}}
+{{- define "consul.chart" -}}
+{{- printf "%s-%s" .Chart.Name .Chart.Version | replace "+" "_" | trunc 63 | trimSuffix "-" -}}
+{{- end -}}
```
[the remark this answers is not in the document]
Comment (tompizmor) [reply incubator/consul/templates/_helpers.tpl]: will do

File: incubator/consul/templates/NOTES.txt (hunk 4 above)
Comment (tompizmor) [reply incubator/consul/templates/NOTES.txt]: ok, I re-phrased it a bit

File: incubator/consul/templates/statefulset.yaml [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+            storage: {{ .Values.persistence.size | quote }}
+      {{- if .Values.persistence.storageClass }}
+      {{- if (eq "-" .Values.persistence.storageClass) }}
+        storageClassName: ""
+      {{- else }}
+        storageClassName: "{{ .Values.persistence.storageClass }}"
+      {{- end }}
+      {{- end }}
+{{- else }}
+        - name: data
```
Comment (sameersbn) [inline incubator/consul/templates/statefulset.yaml]: this volume is not being used if `persistence.enabled=false`

File: incubator/consul/templates/statefulset.yaml [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+          emptyDir: {}
+{{- end }}
+
+
```
Comment (sameersbn) [inline incubator/consul/templates/statefulset.yaml]: remove extra new lines

File: incubator/consul/templates/ui-service.yaml [hunk 8]
```diff
@@ -0,0 +1,22 @@
+{{- if .Values.uiService.enabled }}
+apiVersion: v1
+kind: Service
+metadata:
+  name: "{{ template "consul.fullname" . }}-ui"
+  labels:
+    app: "{{ template "consul.name" . }}"
+    chart: {{ template "consul.chart" . }}
+    release: {{ .Release.Name | quote }}
+    heritage: {{ .Release.Service | quote }}
```
Comment (sameersbn) [inline incubator/consul/templates/ui-service.yaml]: should we be adding a label `component: "ui"` to the UI resources?

File: incubator/consul/templates/ui-service.yaml [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+spec:
+  ports:
+  - name: http
+    port: 80
+    targetPort: http
+  type: "{{ .Values.uiService.type }}"
+  selector:
```
Comment (sameersbn) [inline incubator/consul/templates/ui-service.yaml]: its sufficient to use `app` and `release` for the selector labels.

File: incubator/consul/values.yaml [hunk 10]
```diff
@@ -0,0 +1,208 @@
+## Bitnami Osclass image version
+## ref: https://hub.docker.com/r/bitnami/osclass/tags/
```
Comment (sameersbn) [inline incubator/consul/values.yaml]: `Osclass`?! Don't think so :slightly_smiling_face:

File: incubator/consul/values.yaml [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+##
+image:
+  registry: docker.io
+  repository: bitnami/consul
+  tag: 1.0.6
+  ## Specify a imagePullPolicy
+  ## Defaults to 'Always' if image tag is 'latest', else set to 'IfNotPresent'
```
Comment (sameersbn) [inline incubator/consul/values.yaml]: Remove this comment

File: incubator/consul/values.yaml [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+
+## Consul replicas
+replicas: 3
+
+## Consul service ports
+httpPort: 8500
+rpcPort: 8400
+serflanPort: 8301
+serverPort: 8300
+consulDnsPort: 8600
```
Comment (sameersbn) [inline incubator/consul/values.yaml]: should we instead add a service key, as such:

```
service:
  type: ClusterIP
  ports:
    http: 8500
    rpc: 8400
    serflan: 8301
    server: 8300
    consulDns: 8600
```

the ui port will have be moved out of here. wdyt?

File: incubator/consul/values.yaml [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+#localConfig: |-
+#  {
+#    "key": "value"
+#  }
+
+## Consul domain name.
+domain: consul
+
+## Consul address in which will bind client interfaces.
+clientAddress: '0.0.0.0'
```
Comment (sameersbn) [inline incubator/consul/values.yaml]: Should we expose these options? Shouldn't it always be `0.0.0.0` (same for `serflanAddress`)?

File: incubator/consul/values.yaml [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+#         - true
+
+## Anti-Affinity setting. The default "hard" will use pod anti-affinity that is
+## requiredDuringSchedulingIgnoredDuringExecution to ensure 2 services don't
+## end up on the same node. Setting this to "soft" will use
+## preferredDuringSchedulingIgnoredDuringExecution. If set to anything else,
+## no anti-affinity rules will be configured.
+antiAffinity: "soft"
+
+
```
Comment (sameersbn) [inline incubator/consul/values.yaml]: remove extra new lines

File: incubator/consul/values.yaml [hunk 15]
```diff
@@ -0,0 +1,208 @@
...
+##
+uiService:
+  enabled: true
+  type: "ClusterIP"
+
+## Configure the ingress resource that allows you to access the
+## Consul user interface. Set up the URL
+## ref: http://kubernetes.io/docs/user-guide/ingress/
+##
+uiIngress:
```
Comment (sameersbn) [inline incubator/consul/values.yaml]: should we be adding a `ui` key for the `service` and `ingress` parameters?

f.e.

```yaml
ui:
  service:
    ...
    ...
  ingress:
    ...
    ...
```

File: incubator/consul/values.yaml [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+#        "server":8300
+#    },
+#    "serf_lan":"0.0.0.0"
+#    }
+
+metrics:
+  enabled: false
+  image: prom/consul-exporter
+  imageTag: v0.3.0
+  imagePullPolicy: IfNotPresent
```
Comment (sameersbn) [inline incubator/consul/values.yaml]: for consistency lets use the new image structure here as well

File: incubator/consul/values.yaml [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+  annotations: {}
+
+## Node selector
+## ref: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#nodeselector
+nodeSelector: {"beta.kubernetes.io/arch": "amd64"}
+
+## Configure extra options for liveness and readiness probes
+## ref: https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/#configure-probes)
+livenessProbe:
+  initialDelaySeconds: 30
```
Comment (sameersbn) [inline incubator/consul/values.yaml]: option to enable/disable the probes would be useful for debugging.

**Comment by tompizmor** [comment]
I messed up my changes 😓
Closing this PR in favor of https://github.com/bitnami/charts/pull/652

**Pull request closed without merging**